src/PostgREST/DbStructure/Proc.hs

```diff
@@ -77,30 +79,39 @@ type ProcsMap = M.HashMap QualifiedIdentifier [ProcDescription]
   An overloaded function can have a different volatility or even a different return type.
   Ideally, handling overloaded functions should be left to pg itself. But we need to know certain proc attributes in advance.
 -}
-findProc :: QualifiedIdentifier -> S.Set Text -> Bool -> ProcsMap -> ProcDescription
-findProc qi payloadKeys paramsAsSingleObject allProcs = fromMaybe fallback bestMatch
+filterProc :: QualifiedIdentifier -> S.Set Text -> Bool -> ProcsMap -> [ProcDescription]
```

Comment:
I think the correct return type here should be an `Either ApiRequestError ProcDescription` and not `[ProcDescription]`. Your work [below](https://github.com/PostgREST/postgrest/pull/1841/files#diff-6f407ba6cfd206036588fcd2d4d9583a65800906f957ecd40d79f9e2fec41bffR316-R318) suggests that should be the case. Also our `findRel` function for finding the relationships has the same type.

Reply:
Let me respond do that, as I suggested earlier to make it return `[ProcDescription]` here:

This is in `dbStructure/Proc.hs` - I think this code should not be concerned about returning `ApiRequestError`s. This helper just returns the list of all matching procs we have in the cache.

The lines you referenced are in `ApiRequest.hs`, where it makes a lot of sense to turn the number of procs returned into `ApiRequestError`s, when <> 1.

---

Although, there are a few bits in this function right now that really belong into `ApiRequest` already: The parsing of the cast `::` syntax for one. But also "payload" in `payloadKeys` and `paramsAsSingleObject` are `ApiRequest` concerns.

It would be better overall, if the `filterProc` would have an interface that takes a list of arguments (name + type) to look for.

Maybe create a type like `data Argument = UnnamedArgument | NamedArgument Text | TypedArgument Text Text` or something like that and then make the function take `[Argument]` instead of `S.Set Text -> Bool`.

Reply:
> This is in dbStructure/Proc.hs - I think this code should not be concerned about returning ApiRequestErrors. 

True, but how about if we move the function to ApiRequest.hs?

I think `findProc` should be similar to [findRel](https://github.com/PostgREST/postgrest/blob/5c75f0dcc295e6bd847af6d9703fad5b9c3d76c9/src/PostgREST/Request/DbRequestBuilder.hs#L151-L165), which has complex logic, but it's all encapsulated in a single function(that also returns the Left or Right). Having that logic split between different functions in different modules would make it harder to understand.

Reply:
Yeah, that would be fine, too. Although the raw parsing of `::` casts should still happen at an earlier stage, I think.